include/psa/crypto.h

```diff
@@ -189,11 +187,11 @@ psa_status_t psa_crypto_init(void);
  *    psa_key_derivation_output_key() or psa_copy_key(). This function reads
  *    the attribute structure, creates a key with these attributes, and
  *    outputs a handle to the newly created key.
- * -# The attribute structure is now no longer necessary. If you called
- *    psa_set_key_domain_parameters() earlier, you must call
- *    psa_reset_key_attributes() to free any resources used by the
- *    domain parameters. Otherwise calling psa_reset_key_attributes()
- *    is optional.
+ * -# The attribute structure is now no longer necessary.
+ *    You may call psa_reset_key_attributes(), although this is optional
+ *    with the workflow presented here because the attributes currently
+ *    defined in this specification do not require any additional resources
+ *    beyond the structure itself.
```

Comment:
I'm not sure we should make this optional. Certain implementations might require `psa_reset_key_attributes()` to be called to free implementation-specific resources like domain parameters. It'd be nice if applications would continue to work when domain parameters become part of the official API.

Reply:
We debated this internally and concluded that we might as well allow it.

Reply:
I’m proposing stronger than allowing it, but to maintain requiring it.

Reply:
See internal discussion https://github.com/ARMmbed/psa-crypto/issues/140#issuecomment-486767289 and https://github.com/ARMmbed/mbedtls-psa/pull/263#discussion_r279832614

Reply:
I've read those threads.

My recommendation is to encourage future-proof use of the API by always requiring a reset.

```diff
-  You may call psa_reset_key_attributes(), although this is optional...
+ You must call psa_reset_key_attributes() to free any resources used by the attributes structure.
```

I can see your chosen approach is to only require calling reset when it might be actually necessary; and without key domain parameters, Mbed Crypto doesn't allocate resources when handling attributes. I think this assumes too much about the implementation, as some implementations could require the user calling reset for the attributes currently specified, although Mbed Crypto wouldn't. Applications written for an implementation that doesn't require calling reset will not be portable to implementations that require reset.

If we dont't want to always require it, I'd suggest removing reset from the API for now, so that no implementations that require it appear and cause application portability issues.

Reply:
I don't have strong feelings about this, but for me the removal seems the best solution. Mainly because I am secretly hoping that we can get away without adding domain parameters back:
- I think we have a reasonable compromise for RSA, it has less significance and it's popularity is decreasing.
- I don't think anybody will mourn DSA
- For FFDH named parameters are the safest (if you are worried about logjam, don't use small modulus) and we can use an ECC-like mechanism as we do it now.
- For PQ schemes to come: fiddling with the domain parameters has usually non-trivial impact on security strength and anything else than named parameters seems impractical to me. (The exception is hash based PQ, but there the parameters are as simple as with symmetric crypto.)

Reply:
I'm also feeling that this halfway house is not ideal. There's still the issue with how to extract a variable sized key attributes in an efficient way when we reintroduce them. I think the real options are:

1. Extended attributes are not part of the psa_key_attributes_t.
   * When we re-introduce domain parameters and other similar 'large' attributes, we create a different type to hold them.
   * If you need to extract them, you call an extra API to retrieve them and then release them later.
   * When creating keys, we either allow the extended attributes to be referenced by the key_attributes (but not owned), or add extra parameters to the creation APIs. (I prefer the former)
   * Retain `reset()` to zero the attributes if apps want to do this, is is not required to be called. Releasing memory is an operation on the extended attributes.

2. Make `reset()` mandatory for applications to use in this version of the spec. When domain params are added to the attributes, `reset()` starts checking if there is memory to be released. This adds cost to all apps as many won't touch the zero'd attributes after reset and many compilers will not elide the pointless memset().

3. Remove `reset()` from the API right now. Reintroduce it as "conditional mandatory" - **if you use domain parameters then you must call `reset()`**

I think I actually prefer approach 1. It moves the tricky memory management into a separate structure and allows `psa_key_attributes_t` to always be a simple plain-old-data struct.

Reply:
The risk of some implementations requiring free to come up for the currentlt defined parameters is fairly low. Reset just zeros in this version, and is optional.

Reply:
With my gatekeeper hat for the API, I say: this stays as is in beta3. It's out of scope of this PR anyway. Let's discuss it further for the next beta: we can still make reset officially mandatory or completely optional. This commits us to keeping the function, but it's useful anyway to zero the structure.